pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,950 @@
+		return reconcile.Result{}, err
+	}
+	resourceHelper, err := r.resourceHelperBuilder(restConfig, logger)
+	if err != nil {
+		log.WithError(err).Error("cannot create helper")
+		return reconcile.Result{}, err
+	}
+
+	needToCreateClusterSync := false
+	clusterSync := &hiveintv1alpha1.ClusterSync{}
+	switch err := r.Get(context.Background(), request.NamespacedName, clusterSync); {
+	case apierrors.IsNotFound(err):
+		logger.Info("ClusterSync does not exist; will need to create")
+		needToCreateClusterSync = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not get ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	needToCreateLease := false
+	leaseKey := client.ObjectKey{
+		Namespace: cd.Namespace,
+		Name:      fmt.Sprintf("%s-sync", cd.Name),
+	}
+	lease := &hiveintv1alpha1.ClusterSyncLease{}
+	switch err := r.Get(context.Background(), leaseKey, lease); {
+	case apierrors.IsNotFound(err):
+		logger.Info("Lease for ClusterSync does not exist; will need to create")
+		needToCreateLease = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create lease for ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	origStatus := clusterSync.Status.DeepCopy()
+
+	syncSets, err := r.getSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	selectorSyncSets, err := r.getSelectorSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	needToDoFullReapply := needToCreateClusterSync || r.timeUntilFullReapply(lease) <= 0
+	if needToDoFullReapply {
+		logger.Info("need to reapply all syncsets")
+	}
+
+	// Apply SyncSets
+	syncStatusesForSyncSets, syncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SyncSet",
+		syncSets,
+		clusterSync.Status.SyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SyncSets = syncStatusesForSyncSets
+
+	// Apply SelectorSyncSets
+	syncStatusesForSelectorSyncSets, selectorSyncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SelectorSyncSet",
+		selectorSyncSets,
+		clusterSync.Status.SelectorSyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SelectorSyncSets = syncStatusesForSelectorSyncSets
+
+	setFailedCondition(clusterSync)
+
+	// Create or Update the ClusterSync
+	if needToCreateClusterSync {
+		logger.Info("creating ClusterSync")
+		clusterSync.Namespace = cd.Namespace
+		clusterSync.Name = cd.Name
+		clusterSync.OwnerReferences = []metav1.OwnerReference{{
+			APIVersion:         hivev1.SchemeGroupVersion.String(),
+			Kind:               "ClusterDeployment",
+			Name:               cd.Name,
+			UID:                cd.UID,
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}}
+		if err := r.Create(context.Background(), clusterSync); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create ClusterSync")
+			return reconcile.Result{}, err
```

Comment:
NTS: These error conditions would be pretty sad-making, since we will have tried (with anywhere from total success to total failure) to apply the resources, but will have no record of it. It should shake out eventually -- a subsequent Reconcile will act like this one never happened and do a full reapply. But in the interim, the cluster state will be something we didn't expect.

Not a whole lot we can do about it, though.